### PR TITLE
Chunks

### PR DESCRIPTION
**Pull request by SkidanovAlex**

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -109,8 +113,13 @@ impl Chain {
...
         // Get runtime initial state and create genesis block out of it.
-        let (state_store_update, state_root) = runtime_adapter.genesis_state(0);
-        let genesis = Block::genesis(state_root, genesis_time);
+        let genesis_states = (0..runtime_adapter.num_shards())
+            .map(|shard_id| runtime_adapter.genesis_state(shard_id))
+            .collect::<Vec<_>>();
+        let genesis_state_roots =
+            genesis_states.iter().map(|(_, state_root)| state_root).cloned().collect::<Vec<_>>();
+        let genesis =
+            Block::genesis(genesis_state_roots, genesis_time, runtime_adapter.num_shards());
```
Comment (ilblackdragon) [inline chain/chain/src/chain.rs]: should we parameterize `num_shards` by block index?

File: chain/chunks/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+
+        Ok(ret)
+    }
+
+    pub fn process_orphaned_one_parts(&mut self, block_hash: CryptoHash) -> bool {
+        let mut ret = false;
+        error!("MOO processing orphaned one parts after receiving {:?}", block_hash);
+        for shard_id in 0..self.runtime_adapter.num_shards() {
+            let shard_id = shard_id as ShardId;
+            for part_id in 0..self.runtime_adapter.num_total_parts(block_hash) {
```
Comment (mfornet) [inline chain/chunks/src/lib.rs]: We should avoid calling `num_total_parts` multiple times.

File: chain/network/src/types.rs [hunk 3]
```diff
@@ -773,11 +955,22 @@ pub enum NetworkRequests {
     /// Request given block headers.
     BlockHeadersRequest { hashes: Vec<CryptoHash>, peer_id: PeerId },
     /// Request state for given shard at given state root.
-    StateRequest { shard_id: ShardId, hash: CryptoHash, peer_id: PeerId },
+    StateRequest { shard_id: ShardId, hash: CryptoHash, account_id: AccountId },
     /// Ban given peer.
     BanPeer { peer_id: PeerId, ban_reason: ReasonForBan },
     /// Announce account
-    AnnounceAccount(AnnounceAccount),
+    AnnounceAccount(AnnounceAccount, bool),
```
Comment (mfornet) [inline chain/network/src/types.rs]: I already handled this case in `routing`. The only situation when we want to force announcement is when we are the originators. Also remove myself from routing table.

File: chain/client/src/client.rs [hunk 4]
```diff
@@ -206,17 +243,40 @@ impl Handler<NetworkClientMessages> for ClientActor {
...
+                                self.block_producer.as_ref().unwrap().account_id,
+                                shard_id
+                            );
+                            self.shards_mgr.insert_transaction(shard_id, valid_transaction);
+                            NetworkClientResponses::ValidTx
+                        }
+                        Err(err) => NetworkClientResponses::InvalidTx(err),
+                    }
+                } else {
+                    NetworkClientResponses::NoResponse
```
Comment (mfornet) [inline chain/client/src/client.rs]: Is it expected that tx get dropped in this case?

Comment (ilblackdragon) [reply chain/client/src/client.rs]: We should probably route it to a validator that is in the shard for this tx.
But if we are doing it already on networking level, here we can just drop - as we are assuming it doesn't arrive here unless it's some edge case (like we just switched the caring shards).

**Pull request merged**